x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/categorization/GrokPatternCreator.java

```diff
@@ -0,0 +1,244 @@
+            new GrokPatternCandidate("EMAILADDRESS", "email"),
+            // TODO: would be nice to have IPORHOST here, but HOST matches almost all words
+            new GrokPatternCandidate("IP", "ipaddress"),
+            // This already includes pre/post break conditions
+            new GrokPatternCandidate("QUOTEDSTRING", "field", "", ""),
+            // Can't use \b as the break before, because it doesn't work for negative numbers (the
+            // minus sign is not a "word" character)
+            new GrokPatternCandidate("NUMBER", "field", "(?<!\\w)"),
+            // Disallow +, - and . before hex numbers, otherwise this pattern will pick up base 10
+            // numbers that NUMBER rejected due to preceeding characters
+            new GrokPatternCandidate("BASE16NUM", "field", "(?<![\\w.+-])")
+            // TODO: also unfortunately can't have USERNAME in the list as it matches too broadly
+            // Fixing these problems with overly broad matches would require some extra intelligence
+            // to be added to remove inappropriate matches.  One idea would be to use a dictionary,
+            // but that doesn't necessarily help as "jay" could be a username but is also a dictionary
+            // word (plus there's the international headache with relying on dictionaries).  Similarly,
+            // hostnames could also be dictionary words - I've worked on machines called "hippo" and
+            // "scarf" in the past.  Another idea would be to look at the adjacent characters and
+            // apply some heuristic based on those.
+    );
+
+    private GrokPatternCreator() {
+    }
+
+    /**
+     * Given a category definition regex and a collection of examples from the category, return
+     * a grok pattern that will match the category and pull out any likely fields.  The extracted
+     * fields are given pretty generic names, but unique within the grok pattern provided.  The
+     * expectation is that a user will adjust the extracted field names based on their domain
+     * knowledge.
+     */
+    public static String findBestGrokMatchFromExamples(String regex, Collection<String> examples) {
+
+        // The first string in this array will end up being the empty string, and it doesn't correspond
+        // to an "in between" bit.  Although it could be removed for "neatness", it actually makes the
+        // loops below slightly neater if it's left in.
+        //
+        // E.g., ".*?cat.+?sat.+?mat.*" -> [ "", "cat", "sat", "mat" ]
+        String[] fixedRegexBits = regex.split("\\.[*+]\\??");
+
+        // Create a pattern that will capture the bits in between the fixed parts of the regex
+        //
+        // E.g., ".*?cat.+?sat.+?mat.*" -> Pattern (.*?)cat(.+?)sat(.+?)mat(.*)
+        Pattern exampleProcessor = Pattern.compile(regex.replaceAll("(\\.[*+]\\??)", "($1)"), Pattern.DOTALL);
+
+        List<Collection<String>> inBetweenBits = new ArrayList<>(fixedRegexBits.length);
```

Comment:
Maybe `groupsMatches` instead of `inBetweenBits` or `matchesForGroups`. This is and array of example matches for each group